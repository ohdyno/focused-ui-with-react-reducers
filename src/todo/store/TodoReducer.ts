import {TodoAction} from "./TodoActions.ts";

export enum TodoStatus {
    Complete = "Complete",
    Incomplete = "Incomplete",
}

export type TodoItem = {
    id: string;
    title: string,
    status: TodoStatus
}

function NewTodoItem(title: string, id: string): TodoItem {
    return {id, title, status: TodoStatus.Incomplete}
}

export type TodoState = {
    todos: TodoItem[]
}

// eslint-disable-next-line @typescript-eslint/no-unused-vars
function nonExhaustiveCaseDetected(_case: never): never {
    throw Error("Non-exhaustive case reached. Typescript compilation step should have failed.")
}

export default function reducer(state: TodoState, action: TodoAction) {
    const actionType = action.type;
    switch (actionType) {
        case "load todos":
            return {...state, todos: action.todos}
        case "new todo":
            if (action.content.trim().length > 0) {
                return {...state, todos: [NewTodoItem(action.content, state.todos.length.toString()), ...state.todos]};
            }
            return state
        case "toggle todo status":
            return {
                ...state, todos: state.todos.map((todo) => {
                    if (todo.id === action.id) {
                        return {
                            ...todo,
                            status: TodoStatus.Complete === todo.status ? TodoStatus.Incomplete : TodoStatus.Complete
                        }
                    }
                    return todo
                })
            }
        default:
            return nonExhaustiveCaseDetected(actionType)
    }
}
