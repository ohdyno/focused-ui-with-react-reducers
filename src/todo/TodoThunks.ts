import {Dispatch} from "react";
import {LoadTodosAction, TodoAction} from "./TodoActions.ts";
import {TodoItem} from "./TodoReducer.ts";

export type Thunk = (dispatch: Dispatch<TodoAction>) => void

export function isThunk(maybeThunk: unknown): maybeThunk is Thunk {
    return typeof maybeThunk === 'function'
}

export const LoadTodosThunk: Thunk = (dispatch) => {
    fetch("https://xingzhou.me/stub-apis/todos")
        .then((response) => response.json())
        .then((todos: TodoItem[]) => {
            dispatch(LoadTodosAction(todos))
        })
}

export function createActionOrThunkDispatch(dispatch: Dispatch<TodoAction>) {
    return (actionOrThunk: TodoAction | Thunk) => {
        if (isThunk(actionOrThunk)) {
            actionOrThunk(dispatch)
            return
        }
        dispatch(actionOrThunk)
    }
}

