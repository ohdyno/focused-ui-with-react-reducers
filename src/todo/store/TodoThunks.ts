import {LoadTodosAction, TodoAction} from "./TodoActions.ts";
import {TodoItem} from "./TodoReducer.ts";
import {Thunk as ReduxThunk} from "../../simple-redux/simple-redux.ts";

type Thunk = ReduxThunk<TodoAction>
export const LoadTodosThunk: Thunk = (dispatch) => {
    fetch("https://xingzhou.me/stub-apis/todos")
        .then((response) => response.json())
        .then((todos: TodoItem[]) => {
            dispatch(LoadTodosAction(todos))
        })
}
