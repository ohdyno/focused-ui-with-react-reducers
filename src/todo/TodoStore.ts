import {createContext, Dispatch, useContext} from "react";
import {TodoState, TodoStatus} from "./TodoReducer.ts";
import {TodoAction} from "./TodoActions.ts";
import {Thunk} from "./TodoThunks.ts";

export type TodoStore = {
    state: TodoState,
    dispatch: Dispatch<TodoAction | Thunk>
}
export const defaultState: TodoState = {
    todos: [
        {
            id: "0",
            title: "mow the lawn",
            status: TodoStatus.Complete
        },
        {
            id: "1",
            title: "feed the dog",
            status: TodoStatus.Incomplete
        }
    ]
};

export const StoreState = createContext<TodoState>(defaultState)

export const defaultDispatch: Dispatch<TodoAction | Thunk> = () => {/* do nothing */
};
export const StoreDispatch = createContext<Dispatch<TodoAction | Thunk>>(defaultDispatch)

export function useDispatch(): Dispatch<TodoAction | Thunk> {
    return useContext(StoreDispatch);
}

export function useStoreState(): TodoState {
    return useContext(StoreState);
}
