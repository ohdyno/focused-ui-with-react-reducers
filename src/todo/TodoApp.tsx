import {FormEvent, PropsWithChildren, useEffect, useState} from "react";
import {useDispatch, useStoreState} from "./store/TodoStore.ts";
import {TodoItem, TodoState, TodoStatus} from "./store/TodoReducer.ts";
import {NewTodoAction, ToggleTodoStatusAction} from "./store/TodoActions.ts";
import {LoadTodosThunk} from "./store/TodoThunks.ts";

function TodoList({todos}: TodoState) {
    const dispatch = useDispatch()

    function handleOnChange(todo: TodoItem) {
        dispatch(ToggleTodoStatusAction(todo))
    }

    return (
        <ul>
            {todos.map((todo) => {
                const id = todo.title
                    .replace(/\s+/g, '-')
                    .toLowerCase()
                return (
                    <li key={id}>
                        <label>
                            {todo.title}
                            <input type={"checkbox"}
                                   checked={todo.status === TodoStatus.Complete}
                                   onChange={() => handleOnChange(todo)}/>
                        </label>
                    </li>
                )
            })}
        </ul>
    )
}

export function TodoApp() {
    const state = useStoreState()
    const dispatch = useDispatch()

    useEffect(() => {
        dispatch(LoadTodosThunk)
    }, [dispatch])

    if (state.todos.length > 0) {
        return (
            <Todo>
                <TodoList todos={state.todos}/>
            </Todo>
        )
    }

    return (
        <Todo>
            <section data-testid={"todo-list"}>
                <h1>All Done!</h1>
            </section>
        </Todo>
    );
}

function Todo({children}: PropsWithChildren) {
    const dispatch = useDispatch();
    const [todo, setTodo] = useState('')

    function handleOnSubmit(e: FormEvent<HTMLFormElement>) {
        e.preventDefault()
        dispatch(NewTodoAction(todo))
        setTodo('')
    }

    return (
        <section>
            <form method={'post'} onSubmit={handleOnSubmit}>
                <label>Add New Todo:
                    <input type={"text"} placeholder={"New Todo"} value={todo}
                           onChange={(event) => setTodo(event.target.value)}/>
                </label>
            </form>
            <section>
                {children}
            </section>
        </section>
    )
}
