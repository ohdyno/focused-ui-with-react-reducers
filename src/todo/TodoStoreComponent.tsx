import {PropsWithChildren} from "react";
import {defaultDispatch, defaultState, StoreDispatch, StoreState, TodoStore} from "./TodoStore.ts";

export function Store({state = defaultState, dispatch = defaultDispatch, children}: PropsWithChildren<Partial<TodoStore>>) {
    return (
        <StoreState.Provider value={state}>
            <StoreDispatch.Provider value={dispatch}>
                {children}
            </StoreDispatch.Provider>
        </StoreState.Provider>
    )
}
