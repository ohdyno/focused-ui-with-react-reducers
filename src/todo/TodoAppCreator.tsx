import {useMemo, useReducer} from "react";
import {defaultState} from "./TodoStore.ts";
import {TodoApp} from "./TodoApp.tsx";
import reducer from "./TodoReducer.ts";
import {createActionOrThunkDispatch} from "./TodoThunks.ts";
import {Store} from "./TodoStoreComponent.tsx";

export function TodoAppCreator() {
    const [state, dispatch] = useReducer(reducer, defaultState);
    const thunkDispatch = useMemo(() => createActionOrThunkDispatch(dispatch), [dispatch]);

    return (
        <Store state={state} dispatch={thunkDispatch}>
            <TodoApp/>
        </Store>
    )
}
