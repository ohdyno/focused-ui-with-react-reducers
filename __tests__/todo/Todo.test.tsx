import {describe, expect, it, vi} from "vitest";
import {render, screen} from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import {TodoApp} from "../../src/todo/TodoApp";
import {TodoStatus} from "../../src/todo/TodoReducer";
import {NewTodoAction, ToggleTodoStatusAction} from "../../src/todo/TodoActions";
import {LoadTodosThunk} from "../../src/todo/TodoThunks";
import {Store} from "../../src/todo/TodoStoreComponent";

describe('Todo', () => {
    describe('Rendering', () => {
        describe('No (Zero) Todos', () => {
            it('renders "All done!"', () => {
                const {container} = render(
                    <Store state={{todos: []}}>
                        <TodoApp/>
                    </Store>
                );

                expect(container).toMatchSnapshot()
            });
        })

        describe('Have Todos', () => {
            it('renders the todo title and status', () => {
                const {container} = render(
                    <Store state={{
                        todos: [
                            {
                                id: "0",
                                title: "mow the lawn",
                                status: TodoStatus.Complete
                            },
                            {
                                id: "1",
                                title: "feed the dog",
                                status: TodoStatus.Incomplete
                            }
                        ]
                    }
                    }>
                        <TodoApp/>
                    </Store>
                );

                expect(container).toMatchSnapshot()
            });
        })
    })

    describe('Handling user interactions', () => {
        describe('Adding a new todo', () => {
            it('dispatches a "new todo" action', async () => {
                const user = userEvent.setup()
                const dispatch = vi.fn()
                render(
                    <Store dispatch={dispatch}>
                        <TodoApp/>
                    </Store>
                );

                const input = screen.getByLabelText(/add new todo/i);
                await user.type(input, "cook food{enter}")

                expect(dispatch).toHaveBeenCalledWith(NewTodoAction("cook food"))
            });

            it('allows inputting multiple todos without deleting previous ', async () => {
                const user = userEvent.setup()
                const dispatch = vi.fn()
                render(
                    <Store dispatch={dispatch}>
                        <TodoApp/>
                    </Store>
                );

                const input = screen.getByLabelText(/add new todo/i);
                await user.type(input, "cook food{enter}")
                expect(dispatch).toHaveBeenCalledWith(NewTodoAction("cook food"))

                await user.type(input, "take out garbage{enter}")
                expect(dispatch).toHaveBeenCalledWith(NewTodoAction("take out garbage"))
            });
        })

        describe('Changing todo status', () => {
            const todo = {
                id: "0",
                title: "feed the dog",
                status: TodoStatus.Incomplete
            };

            it('dispatches a "toggle todo status" action when a todo status checkbox is checked/unchecked', async () => {
                const user = userEvent.setup()
                const dispatch = vi.fn()
                render(
                    <Store state={{todos: [todo]}}
                           dispatch={dispatch}>
                        <TodoApp/>
                    </Store>
                );

                const checkBox = screen.getByLabelText(new RegExp(todo.title, "i"));
                await user.click(checkBox)

                expect(dispatch).toHaveBeenCalledWith(ToggleTodoStatusAction(todo))
            });
        })
    })

    describe('Loading todos on initial render', () => {
        it('dispatches the "load todos" thunk', () => {
            const dispatch = vi.fn()
            render(
                <Store dispatch={dispatch}>
                    <TodoApp/>
                </Store>
            );
            expect(dispatch).toHaveBeenCalledWith(LoadTodosThunk)
        })
    })
});
