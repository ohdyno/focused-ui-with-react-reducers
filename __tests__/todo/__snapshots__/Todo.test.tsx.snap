// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Todo > Rendering > Have Todos > renders the todo title and status 1`] = `
<div>
  <section>
    <form
      method="post"
    >
      <label>
        Add New Todo:
        <input
          placeholder="New Todo"
          type="text"
          value=""
        />
      </label>
    </form>
    <section>
      <ul>
        <li>
          <label>
            mow the lawn
            <input
              checked=""
              type="checkbox"
            />
          </label>
        </li>
        <li>
          <label>
            feed the dog
            <input
              type="checkbox"
            />
          </label>
        </li>
      </ul>
    </section>
  </section>
</div>
`;

exports[`Todo > Rendering > No (Zero) Todos > renders "All done!" 1`] = `
<div>
  <section>
    <form
      method="post"
    >
      <label>
        Add New Todo:
        <input
          placeholder="New Todo"
          type="text"
          value=""
        />
      </label>
    </form>
    <section>
      <section
        data-testid="todo-list"
      >
        <h1>
          All Done!
        </h1>
      </section>
    </section>
  </section>
</div>
`;
